File: system-test/dotnet/Driver.Adapter/External/Tax/Client/BaseTaxClient.cs
using Driver.Adapter.Shared.Client.Http;
using Common;
using Driver.Adapter.External.Tax.Client.Dtos;
using Driver.Adapter.External.Tax.Client.Dtos.Error;

namespace Driver.Adapter.External.Tax.Client;

public abstract class BaseTaxClient : IDisposable
{
    private const string HealthEndpoint = "/health";
    private const string CountriesEndpoint = "/api/countries";

    protected readonly JsonHttpClient<ExtTaxErrorResponse> _httpClient;
    private bool _disposed;

    protected BaseTaxClient(string baseUrl)
    {
        _httpClient = new JsonHttpClient<ExtTaxErrorResponse>(baseUrl);
    }

    public void Dispose()
    {
        Dispose(true);
        GC.SuppressFinalize(this);
    }

    protected virtual void Dispose(bool disposing)
    {
        if (_disposed) return;
        if (disposing)
            _httpClient?.Dispose();
        _disposed = true;
    }

    public Task<Result<VoidValue, ExtTaxErrorResponse>> CheckHealthAsync()
        => _httpClient.GetAsync(HealthEndpoint);

    public Task<Result<ExtCountryDetailsResponse, ExtTaxErrorResponse>> GetCountryAsync(string? country)
        => _httpClient.GetAsync<ExtCountryDetailsResponse>($"{CountriesEndpoint}/{country}");
}

File: system-test/java/src/main/java/com/mycompany/myshop/testkit/driver/adapter/external/tax/client/BaseTaxClient.java
package com.mycompany.myshop.testkit.driver.adapter.external.tax.client;

import com.mycompany.myshop.testkit.driver.adapter.external.tax.client.dtos.ExtGetCountryResponse;
import com.mycompany.myshop.testkit.driver.adapter.external.tax.client.dtos.error.ExtTaxErrorResponse;
import com.mycompany.myshop.testkit.driver.adapter.shared.client.http.JsonHttpClient;
import com.mycompany.myshop.testkit.common.Closer;
import com.mycompany.myshop.testkit.common.Result;

public abstract class BaseTaxClient implements AutoCloseable {
    private static final String HEALTH_ENDPOINT = "/health";
    private static final String COUNTRIES_ENDPOINT = "/api/countries/";

    protected final JsonHttpClient<ExtTaxErrorResponse> httpClient;

    protected BaseTaxClient(String baseUrl) {
        this.httpClient = new JsonHttpClient<>(baseUrl, ExtTaxErrorResponse.class);
    }

    @Override
    public void close() {
        Closer.close(httpClient);
    }

    public Result<Void, ExtTaxErrorResponse> checkHealth() {
        return httpClient.get(HEALTH_ENDPOINT);
    }

    public Result<ExtGetCountryResponse, ExtTaxErrorResponse> getCountry(String country) {
        return httpClient.get(COUNTRIES_ENDPOINT + country, ExtGetCountryResponse.class);
    }
}
